kOS/pegas_util.ks

```diff
@@ -735,6 +743,56 @@ FUNCTION stageEventHandler {
 	setNextEvent(currentTime, eventDelay).
 }.
 
+//	Handles communication with other CPUs
+FUNCTION commsEventHandler {
+	LOCAL message IS CORE:MESSAGES:PEEK:CONTENT.	//	Get the oldest message in the queue but don't delete it yet
+
+	IF message["type"] = "request" {
+
+		//	Make sure data is in a list
+		LOCAL data IS message["data"].
+		IF NOT data:ISTYPE("List") {
+			SET data TO LIST(data).
+		}
+
+		//	List of available return data
+		LOCAL availableData IS LEXICON(
+			"liftOffTime", liftOffTime:SECONDS,
+			"launchAzimuth", mission["launchAzimuth"],
+			"upfgConverged", upfgConverged
+		).
+
+		//	Create a response to send back
+		LOCAL response IS LEXICON("type", "response").
+
+		//	Iterate through requested data and add to response
+		FOR d IN data {
+			IF d:ISTYPE("String") {	//	Only do the following if data is a string
+				IF availableData:HASKEY(d) {
+					response:ADD(d, availableData[d]).
+				}
+			}
+		}
+
+		//	Add the sender info - optional
+		//	kOS Message only reports the sender vessel (don't know why), not CPU so this fixes that.
+		response:ADD("sender", CORE:TAG).
+
+		//	Send the response
+		IF message:HASKEY("sender") {
+			PROCESSOR(message["sender"]):CONNECTION:SENDMESSAGE(response).
+		} ELSE {
+			CORE:MESSAGES:PEEK:SENDER:CONNECTION:SENDMESSAGE(response).
+		}
+
+	} ELSE IF message["type"] = "command" {
+		//	TO BE CREATED
```

Comment:
On your [comment](https://github.com/Noiredd/PEGAS/pull/14#issuecomment-337448288) - yes, I think we should do it the same way: list available commands (btw, I like the idea of putting delegates into the lexicon) and have the handler call whatever value it finds under the received key.  
About what the command should return - I think checking whether a command executed properly is not very realistic (how do we make sure of that?). Instead, I would see each delegate returning some value (from simple true/false to more detailed information) which we would then pack into the reply, and have it up to the sender to figure out what to do with them. This reiterates the need for documentation.

On the other hand, how do you feel about the sender being able (or not) to send multiple commands (similar to multiple requests)?

Reply:
I guess multiple commands could be useful in some situations so we could add that.
If we were to have only one command per message, the message could include:
`... "data", LIST("engineShutdown", "engine_1_tag", "engine_2_tag") ...`
So a list with first item being the command name, and the remaining items being parameters.
For multiple commands per message, it could be done like this:
`... "data", LEXICON("engineShutdown", LIST("engine_1_tag", "engine_2_tag")) ...`
So a lexicon with command name as key and list of parameters as values. This could have multiple entries.

I'd be more inclined towards the second option with multiple commands, what do you think?

Reply:
Are lexicons order-preserving? I'm used to Python's `dict`s and C++'s `map`s, which are not - and if we allow the user to send multiple commands, we should make sure that they will be executed in the exact order the user passes them. For this reason I would lean towards lists here. The `data` key would contain a list of lists in the first format:  
`"data", LIST(LIST("engineShutdown", "tag_1", "tag_2"), LIST("setVar", "asd", 130), ...)`  
For ease of use we should do like in the request message, allow omitting the outermost list and have the handler ensure the format adheres to the standard before running the loop.

Reply:
I don't think I ever noticed a lexicon changing order but I made it the way you suggested.